File: backend/dwf/api/routers/workflows.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from dwf.api.deps import get_current_active_user, get_db
from dwf.domain.models.workflow import WorkflowCreate, WorkflowResponse, WorkflowUpdate
from dwf.infrastructure.database.models import User
from dwf.services.workflow_service import WorkflowService

router = APIRouter(prefix="/workflows", tags=["workflows"])


@router.post("/", status_code=status.HTTP_201_CREATED, response_model=WorkflowResponse)
async def create_workflow(
    data: WorkflowCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    service = WorkflowService(db)
    try:
        workflow = await service.create_workflow(data, owner_id=str(current_user.id))
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    return workflow


@router.get("/", response_model=list[WorkflowResponse])
async def list_workflows(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    service = WorkflowService(db)
    return await service.list_by_owner(owner_id=str(current_user.id))


@router.get("/{workflow_id}", response_model=WorkflowResponse)
async def get_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    service = WorkflowService(db)
    workflow = await service.get_by_id(workflow_id, owner_id=str(current_user.id))
    if workflow is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found"
        )
    return workflow


@router.put("/{workflow_id}", response_model=WorkflowResponse)
async def update_workflow(
    workflow_id: str,
    data: WorkflowUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    service = WorkflowService(db)
    workflow = await service.update_workflow(
        workflow_id, owner_id=str(current_user.id), data=data
    )
    if workflow is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found"
        )
    return workflow


@router.delete("/{workflow_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workflow(
    workflow_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    service = WorkflowService(db)
    success = await service.delete_workflow(workflow_id, owner_id=str(current_user.id))
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found"
        )

File: frontend/src/api/workflows.ts
import api from './client';

export interface Workflow {
  id: string;
  name: string;
  description: string | null;
  status: 'draft' | 'active' | 'paused' | 'completed' | 'failed';
  owner_id: string;
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
}

export interface WorkflowUpdate {
  name?: string;
  description?: string;
  status?: Workflow['status'];
}

export const workflowApi = {
  list() {
    return api.get<Workflow[]>('/workflows/');
  },
  get(id: string) {
    return api.get<Workflow>(`/workflows/${id}`);
  },
  create(data: WorkflowCreate) {
    return api.post<Workflow>('/workflows/', data);
  },
  update(id: string, data: WorkflowUpdate) {
    return api.put<Workflow>(`/workflows/${id}`, data);
  },
  delete(id: string) {
    return api.delete(`/workflows/${id}`);
  },
};
